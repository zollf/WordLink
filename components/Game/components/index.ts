export { default as Word } from './Word';
export { default as Letter } from './Letter';
export { default as LetterSelection } from './LetterSelection';
export { default as CurrentWord } from './CurrentWord';
export { default as Path } from './Path';
export { default as ControlButtons } from './ControlButtons';
export { default as Completed } from './Completed';