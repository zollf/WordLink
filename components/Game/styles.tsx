import { StyleSheet } from "react-native";
import { heightPercentageToDP } from 'react-native-responsive-screen';

const styles = StyleSheet.create({
  game: {
    width: "100%",
    height: "100%",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  information: {
    position: "absolute",
    top: heightPercentageToDP('8%'),
  },
  content: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    marginBottom: 80,
  },
  letterSelection: {
    height: 150,
    position: "absolute",
    bottom: heightPercentageToDP('15%'),
  },
  completed: {
    position: "absolute",
    bottom: heightPercentageToDP('15%'),
  },
});

export default styles;