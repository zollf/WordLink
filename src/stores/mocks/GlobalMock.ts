const GlobalMock: Global = {
  _ready: true,
  currentPage: 'index',
  setCurrentPage: jest.fn(),
  previousPage: 'index',
  currentDifficultyOpen: '',
  setCurrentDifficultyOpen: jest.fn((d: string) => (GlobalMock.currentDifficultyOpen = d)),
  getDifficultyColor: jest.fn(() => ''),
  setStartingPage: jest.fn(),
  onStartingPage: false,
  setUserInfo: jest.fn(),
  userInfo: {
    username: 'jeff',
    points: 20,
    puzzles: {
      completed: 2,
      easy: [],
      medium: [],
      hard: [],
    },
    hintsUsed: 6,
    hardestDifficulty: 'Medium',
    starsEarnt: 69,
  },
  init: jest.fn(),
};

export default GlobalMock;
