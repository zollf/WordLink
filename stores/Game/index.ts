import { observable, action, makeObservable } from "mobx";
import { dictionary, getLevel } from '../../data';

class GameStore {
  gameId = 0;
  difficulty = '';
  inGame = false;
  loaded = false;
  error = false;
  game?: Game;
  
  currentWord = '';
  selectedLetter = '';
  stepNumber = 0;
  path: string[] = [];
  visited = {};

  constructor() {
    makeObservable(this, {
      inGame:             observable,
      gameId:             observable,
      path:               observable,
      difficulty:         observable,
      selectedLetter:     observable,
      currentWord:        observable,
      stepNumber:         observable,
      loaded:             observable,
      error:              observable,

      getNewLetters:      action,
      isWord:             action,
      loadGame:           action,
      startGame:          action,
      setSelectedLetter:  action,
      changeLetter:       action,
      resetCurrentWord:   action,
      undo:               action,
      clearGame:          action,
    });
  }

  /**
   * Loads the game file into the store
   * @param gameId 
   * @param difficulty 
   */
  loadGame = (gameId: number, difficulty: string) => {
    this.gameId = gameId;
    this.difficulty = difficulty;

    this.game = getLevel(difficulty)[gameId].game;

    if (this.game && this.game.start && this.game.end) {
      this.loaded = true;
    } else {
      this.error = true;
    }
  };

  /**
   * Starts the game
   */
  startGame = () => {
    if (this.game) {
      this.inGame = true;
      this.currentWord = this.game.start;
      this.path.push(this.currentWord);
      this.visited[this.currentWord] = true;
    } else {
      // log error
    }
  };

  /**
   * Finds all letters from the branching words
   * @param word , the word we want to branch off
   * @return letters , of all the new branching words
   */
  getNewLetters = (): string[] => {
    const letters: string[] = [];
    const wordArr = this.currentWord.split('');
    wordArr.forEach((_, i: number): void => {
      for (let j = 0; j < 26; j++) {
        const newLetter = String.fromCharCode(j + 65);
        const temp = wordArr[i];
        wordArr[i] = newLetter;
        if (this.isWord(wordArr.join('')) && !letters.includes(newLetter)) {
          letters.push(newLetter.toUpperCase());
        }
        wordArr[i] = temp;
      }
    });
    return letters;
  };

  /**
   * Checks if a word is a valid word
   * @param word , word to check if it is within the dictionary 
   * @return if the word is included in the dictionary
   */
  isWord = (word: string): boolean => {
    return !!dictionary[word.toUpperCase()];
  };

  setSelectedLetter = (letter: string) => {
    if (this.selectedLetter === letter) {
      this.selectedLetter = '';
    } else {
      this.selectedLetter = letter;
    }
  };


  /**
   * Change the letter from the current selected letter
   * @param index 
   */
  changeLetter = (index: number) => {
    if (this.selectedLetter) {
      const newWordArr = this.currentWord.split("");
      newWordArr[index] = this.selectedLetter;
      const newWord = newWordArr.join("");

      // The new word is a valid word and has not been visited
      if (this.isWord(newWord) && !this.visited[newWord]) {
        // Add to path
        this.path.push(newWord);
        this.stepNumber++;

        // Replace Current word with the newly made word
        this.currentWord = newWord;

        // Reset current selected letter
        this.selectedLetter =  '';

        // Add to visited
        this.visited[newWord] = true;
      }
    }
  };

  /**
   * Resets the word back to starting
   */
  resetCurrentWord = () => {
    this.stepNumber = 0;
    this.currentWord = this.path[this.stepNumber];
    this.path = [this.currentWord];
    this.visited = {};
  };

  /**
   * Undo the current step, move back one in the path
   */
  undo = () => {
    if (this.stepNumber > 0) {
      this.visited[this.path[this.stepNumber]] = false; 
      this.stepNumber--;
      this.currentWord = this.path[this.stepNumber];
      this.path.pop();
    }
  };

  /**
   * Clears the game, force reset
   */
  clearGame = () => {
    this.inGame = false;
    this.path = [];
    this.loaded = false;
    this.error = false;
    this.stepNumber = 0;
    this.visited = {};
  };
}

const gameStore = new GameStore();
export { GameStore };
export default gameStore;