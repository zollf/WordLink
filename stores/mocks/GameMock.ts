const level: Level = {
  title: 'hello',
  difficulty: 'easy',
  game: {
    start: 'FORTY',
    end: 'FIFTY',
    path: 5,
  },
};

const GameMocks = {
  gameId: 0,
  difficulty: 'EASY',
  inGame: false,
  loaded: false,
  error: false,
  game: level.game,
  gameTitle: 'hello',

  currentWord: level.game.start,
  selectedLetter: '',
  stepNumber: 0,
  path: [level.game.start],
  visited: { FORTY: true },
  completed: false,

  loadGame: jest.fn(),
  startGame: jest.fn(),
  getNewLetters: jest.fn(() => ['A', 'B', 'C']),
  isWord: jest.fn(),
  setSelectedLetter: jest.fn(),
  changeLetter: jest.fn(),
  resetCurrentWord: jest.fn(),
  undo: jest.fn(),
  clearGame: jest.fn(),
};

export default GameMocks;
