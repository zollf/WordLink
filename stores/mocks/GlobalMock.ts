const GlobalMock: Global = {
  currentPage: 'index',
  setCurrentPage: jest.fn(),
  previousPage: 'index',
  currentDifficultyOpen: '',
  setCurrentDifficultyOpen: jest.fn((d: string) => (GlobalMock.currentDifficultyOpen = d)),
  getDifficultyColor: jest.fn(() => ''),
  setStartingPage: jest.fn(),
  onStartingPage: false,
};

export default GlobalMock;
