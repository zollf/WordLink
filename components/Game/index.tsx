import React, { useState, useEffect } from 'react';
import { View, Animated } from 'react-native';
import { observer } from 'mobx-react';
import { useStore, useAnimate } from '../../hooks';
import CircleButton from '../CircleButton';

import { 
  Word, 
  LetterSelection, 
  CurrentWord, 
  ControlButtons, 
  Path,
  Completed,
  ExitModal,
  Information
} from './components';

import { ElasticSlideIn, FadeSlideIn, Animate } from '../../animations';
import styles from './styles';

const Game = () => {
  const { game, clearGame, completed } = useStore('gameStore');
  const { setCurrentPage } = useStore('global');
  const [exitModal, setExitModal] = useState(false);

  const [infoT, infoO] = useAnimate(-100, 0);
  const [word1T, word1O] = useAnimate(-100, 0);
  const [pathT, pathO] = useAnimate(-100, 0);
  const [currentT, currentO] = useAnimate(-100, 0);
  const [word2T, word2O] = useAnimate(-100, 0);
  const [lettersT, lettersO] = useAnimate(-100, 0);
  useEffect(() => {
    Animated.parallel([
      Animated.stagger(20, [
        ElasticSlideIn(infoT),
        ElasticSlideIn(word1T),
        ElasticSlideIn(pathT),
        ElasticSlideIn(currentT),
        ElasticSlideIn(word2T),
        ElasticSlideIn(lettersT),
      ]),
      Animated.stagger(20, [
        FadeSlideIn(infoO),
        FadeSlideIn(word1O),
        FadeSlideIn(pathO),
        FadeSlideIn(currentO),
        FadeSlideIn(word2O),
        FadeSlideIn(lettersO),
      ]),
    ]).start();
  });

  const handleExit = () => {
    setCurrentPage('menu');
    clearGame();
  };

  return (
    <View style={styles.game}>
      {exitModal && <ExitModal close={() => setExitModal(false)} exit={handleExit}/>}
      <View style={styles.back}>
        <CircleButton overrideCallback={() => setExitModal(true)}/>
      </View>
      
      
      <Animate style={styles.information} transform={infoT} opacity={infoO}>
        <Information/>
      </Animate>

      <View style={styles.content}>
        <Animate transform={word1T} opacity={word1O}>
          <Word color="orange" word={game.start} />
        </Animate>
        
        <Animate transform={pathT} opacity={pathO}>
          <Path />
        </Animate>
        
        <Animate transform={currentT} opacity={currentO}>
          <CurrentWord />
        </Animate>

        <Animate transform={word2T} opacity={word2O}>
          <Word color="blue" word={game.end} />
        </Animate>
      </View>

      <Animate style={styles.letterSelection} transform={lettersT} opacity={lettersO}>
        <LetterSelection />
      </Animate>

      {!completed ? (
        <ControlButtons /> 
      ) : ( 
        <View style={styles.completed}>
          <Completed />
        </View>
      )}
    </View>
  );
};

export default observer(Game);