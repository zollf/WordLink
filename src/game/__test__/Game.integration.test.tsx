import React from 'react';
import { Provider } from 'mobx-react';
import GlobalMock from '../../../stores/mocks/GlobalMock';
import GameMock from '../../../stores/mocks/GameMocks';

import Game from '../';

import renderer from 'react-test-renderer';

const component = (
  <Provider global={GlobalMock} gameStore={GameMock}>
    <Game />
  </Provider>
);

describe('<Game />', () => {
  const wrapper = renderer.create(component);
  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('loaded game matches snapshot', () => {
    wrapper.update(component);

    expect(wrapper).toMatchSnapshot();
  });
});