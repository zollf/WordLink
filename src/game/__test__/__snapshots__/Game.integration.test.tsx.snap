// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> loaded game matches snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#40BAFF",
          "borderRadius": 100,
          "display": "flex",
          "height": 35,
          "justifyContent": "center",
          "left": 20,
          "opacity": 1,
          "position": "absolute",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 4,
          "top": 60,
          "width": 35,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../images/left-arrow.png",
          }
        }
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          undefined,
          Object {
            "opacity": mockConstructor {},
            "transform": Array [
              Object {
                "translateX": mockConstructor {},
              },
            ],
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "Raleway_400Regular",
            "fontSize": 24,
          }
        }
      >
        Link the two words
      </Text>
    </View>
    <View
      style={
        Array [
          undefined,
          Object {
            "opacity": mockConstructor {},
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F3F3F3",
            "display": "flex",
            "justifyContent": "space-between",
            "margin": 24,
            "padding": 15,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 4,
            "width": 200,
          }
        }
      >
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "translateY": mockConstructor {},
                },
              ],
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 5,
                  "display": "flex",
                  "height": 60,
                  "justifyContent": "center",
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 2,
                    "width": 2,
                  },
                  "shadowOpacity": 0.1,
                  "shadowRadius": 4,
                  "width": 180,
                },
                Object {
                  "backgroundColor": "#FF9C40",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "fontFamily": "Raleway_400Regular",
                  "fontSize": 18,
                }
              }
            >
              FORTY
            </Text>
          </View>
        </View>
        <Image
          source={
            Object {
              "testUri": "../../../images/transfer.png",
            }
          }
          style={
            Object {
              "height": 40,
              "margin": 20,
              "transform": Array [
                Object {
                  "rotate": "90deg",
                },
              ],
              "width": 40,
              "zIndex": 1,
            }
          }
        />
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "translateY": mockConstructor {},
                },
              ],
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 5,
                  "display": "flex",
                  "height": 60,
                  "justifyContent": "center",
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 2,
                    "width": 2,
                  },
                  "shadowOpacity": 0.1,
                  "shadowRadius": 4,
                  "width": 180,
                },
                Object {
                  "backgroundColor": "#40BAFF",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "fontFamily": "Raleway_400Regular",
                  "fontSize": 18,
                }
              }
            >
              FIFTY
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          undefined,
          Object {
            "opacity": mockConstructor {},
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          },
        ]
      }
    >
      <View
        style={
          Object {
            "transform": Array [
              Object {
                "scale": mockConstructor {},
              },
            ],
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FF9C40",
              "borderRadius": 5,
              "display": "flex",
              "height": 76,
              "justifyContent": "center",
              "opacity": 1,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 2,
                "width": 2,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 4,
              "width": 175,
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "Raleway_400Regular",
                "fontSize": 18,
              }
            }
          >
            START
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Game /> matches snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
/>
`;
